nemo/collections/nlp/models/nlp_model.py

```diff
@@ -80,44 +81,66 @@ def register_bert_model(self):
                     f'Registering BERT model config for {self.bert_model} is not yet supported. Please override this method if needed.'
                 )
 
-    def _setup_tokenizer(self, cfg: DictConfig):
+    def setup_tokenizer(self, cfg: DictConfig):
         """Instantiates tokenizer based on config and registers tokenizer artifacts.
 
         Args:
             cfg (DictConfig): Tokenizer config
         """
+        if self._is_model_being_restored() and os.path.exists('tokenizer.vocab_file'):
+            # model is being restored from .nemo file so tokenizer.vocab_file has precedence
+            vocab_file = self.register_artifact(config_path='tokenizer.vocab_file', src='tokenizer.vocab_file')
+        elif cfg.vocab_file is not None:
+            # use vocab file from config
+            vocab_file = self.register_artifact(config_path='tokenizer.vocab_file', src=cfg.vocab_file)
+        else:
+            vocab_file = None
+
         tokenizer = get_tokenizer(
             tokenizer_name=cfg.tokenizer_name,
-            vocab_file=cfg.vocab_file,
+            vocab_file=vocab_file,
             special_tokens=OmegaConf.to_container(cfg.special_tokens) if cfg.special_tokens else None,
-            tokenizer_model=cfg.tokenizer_model,
+            tokenizer_model=self.register_artifact(config_path='tokenizer.tokenizer_model', src=cfg.tokenizer_model),
         )
         self.tokenizer = tokenizer
-        self.register_tokenizer(cfg)
+
+        if vocab_file is None:
+            # when there is no vocab file we try to get the vocab from the tokenizer and register it
+            self._register_vocab_from_tokenizer(config_path='tokenizer.vocab_file')
 
     @rank_zero_only
```

Comment:
why rank_zero_only here?

Reply:
I don't want multiple processes writing to the same file.